### ISSUE TITLE
`option-if-let-else`: no method named `map_or` found for raw pointer `*const std::option::Option<i32>`

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#15379]
Using the following flags

`--force-warn clippy::option-if-let-else`

this code:

```rust
// unit-test: EarlyOtherwiseBranch

// Tests various cases that the `early_otherwise_branch` opt should *not* optimize

// From #78496
enum E<'a> {
    None,
    Some(&'a E<'a>),
}

// EMIT_MIR early_otherwise_branch_soundness.no_downcast.EarlyOtherwiseBranch.diff
fn no_downcast(e: &E) -> u32 {
    if let E::Some(E::Some(_)) = e { 1 } else { 2 }
}

// SAFETY: if `a` is `Some`, `b` must point to a valid, initialized value
// EMIT_MIR early_otherwise_branch_soundness.no_deref_ptr.EarlyOtherwiseBranch.diff
unsafe fn no_deref_ptr(a: Option<i32>, b: *const Option<i32>) -> i32 {
    match a {
        // `*b` being correct depends on `a == Some(_)`
        Some(_) => match *b {
            Some(v) => v,
            _ => 0,
        },
        _ => 0,
    }
}

fn main() {
    no_downcast(&E::None);
    unsafe { no_deref_ptr(None, std::ptr::null()) };
}

```

caused the following diagnostics:

```
    Checking _9e6149148c30446da0f840babc1a1100f927d254 v0.1.0 (/tmp/icemaker_global_tempdir.f0JXFBTcO4io/icemaker_clippyfix_tempdir.hZ8fQUQCHGtu/_9e6149148c30446da0f840babc1a1100f927d254)
warning: use Option::map_or instead of an if let/else
  --> src/main.rs:21:20
   |
21 |           Some(_) => match *b {
   |  ____________________^
22 | |             Some(v) => v,
23 | |             _ => 0,
24 | |         },
   | |_________^ help: try: `b.map_or(0, |v| v)`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#option_if_let_else
   = note: requested on the command line with `--force-warn clippy::option-if-let-else`

warning: `_9e6149148c30446da0f840babc1a1100f927d254` (bin "_9e6149148c30446da0f840babc1a1100f927d254") generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.41s

```

However after applying these diagnostics, the resulting code:

```rust
// unit-test: EarlyOtherwiseBranch

// Tests various cases that the `early_otherwise_branch` opt should *not* optimize

// From #78496
enum E<'a> {
    None,
    Some(&'a E<'a>),
}

// EMIT_MIR early_otherwise_branch_soundness.no_downcast.EarlyOtherwiseBranch.diff
fn no_downcast(e: &E) -> u32 {
    if let E::Some(E::Some(_)) = e { 1 } else { 2 }
}

// SAFETY: if `a` is `Some`, `b` must point to a valid, initialized value
// EMIT_MIR early_otherwise_branch_soundness.no_deref_ptr.EarlyOtherwiseBranch.diff
unsafe fn no_deref_ptr(a: Option<i32>, b: *const Option<i32>) -> i32 {
    match a {
        // `*b` being correct depends on `a == Some(_)`
        Some(_) => b.map_or(0, |v| v),
        _ => 0,
    }
}

fn main() {
    no_downcast(&E::None);
    unsafe { no_deref_ptr(None, std::ptr::null()) };
}

```
no longer compiled:

```
    Checking _9e6149148c30446da0f840babc1a1100f927d254 v0.1.0 (/tmp/icemaker_global_tempdir.f0JXFBTcO4io/icemaker_clippyfix_tempdir.hZ8fQUQCHGtu/_9e6149148c30446da0f840babc1a1100f927d254)
error[E0599]: no method named `map_or` found for raw pointer `*const std::option::Option<i32>` in the current scope
   --> src/main.rs:21:22
    |
 21 |         Some(_) => b.map_or(0, |v| v),
    |                      ^^^^^^
    |
help: there is a method `map_addr` with a similar name, but with different arguments
   --> /home/matthias/.rustup/toolchains/master/lib/rustlib/src/rust/library/core/src/ptr/const_ptr.rs:244:5
    |
244 |     pub fn map_addr(self, f: impl FnOnce(usize) -> usize) -> Self {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0599`.
error: could not compile `_9e6149148c30446da0f840babc1a1100f927d254` (bin "_9e6149148c30446da0f840babc1a1100f927d254" test) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `_9e6149148c30446da0f840babc1a1100f927d254` (bin "_9e6149148c30446da0f840babc1a1100f927d254") due to 1 previous error

```

Version:
```
rustc 1.90.0-nightly (cc0a5b730 2025-07-31)
binary: rustc
commit-hash: cc0a5b73053c62a3df5f84b3ee85079c9b65fa87
commit-date: 2025-07-31
host: x86_64-unknown-linux-gnu
release: 1.90.0-nightly
LLVM version: 20.1.8

```
